### PR TITLE
Filter table columns in bulk

### PR DESCRIPTION
**Pull request by findepi**
Refactor information_schema.columns handling code.

### CODE REVIEW

**Review by electrum — CHANGES_REQUESTED** [review]
We should deprecate the old method

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/ReadOnlySystemAccessControl.java [hunk 1]
```diff
@@ -153,6 +153,12 @@ public Set<String> filterColumns(SystemSecurityContext context, CatalogSchemaTab
         return columns;
     }
 
+    @Override
+    public Map<SchemaTableName, Set<String>> filterColumns(SystemSecurityContext context, String catalogName, Map<SchemaTableName, Set<String>> tableColumns)
```
Comment (kokosing) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/ReadOnlySystemAccessControl.java:157]: Should we deprecate the other `filterColumns` method?

File: lib/trino-array/pom.xml [hunk 2]
```diff
@@ -13,7 +13,6 @@
 
     <properties>
         <air.main.basedir>${project.parent.basedir}</air.main.basedir>
-        <air.compiler.fail-warnings>true</air.compiler.fail-warnings>
```
Comment (kokosing) [inline lib/trino-array/pom.xml:16]: Separate PR?

**Review by kokosing — APPROVED** [review]

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/ReadOnlySystemAccessControl.java (hunk 1 above)
Comment (findepi) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/ReadOnlySystemAccessControl.java:157]: done

File: lib/trino-array/pom.xml (hunk 2 above)
Comment (findepi) [reply lib/trino-array/pom.xml:16]: sure thing!

Comment (findepi) [reply lib/trino-array/pom.xml:16]: done! -- https://github.com/trinodb/trino/pull/18971

File: core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java [hunk 3]
```diff
@@ -248,61 +248,76 @@ private static Map<SchemaTableName, List<ColumnMetadata>> doListTableColumns(Ses
...
-                            .collect(toImmutableSet()));
-            result.put(
-                    table,
-                    columns.stream()
-                            .filter(column -> allowedColumns.contains(column.getName()))
-                            .collect(toImmutableList()));
-        });
+                    List<ColumnMetadata> columns = metadata.getTableMetadata(session, targetTableHandle).getColumns();
+
+                    Set<String> allowedColumns = accessControl.filterColumns(
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java:305]: Can all redirected tables be collected and batched in a single `accessControl.filterColumns` call?

File: core/trino-main/src/main/java/io/trino/security/AccessControlManager.java [hunk 4]
```diff
@@ -635,6 +636,34 @@ public Set<String> filterColumns(SecurityContext securityContext, CatalogSchemaT
...
 
+    @Override
+    public Map<SchemaTableName, Set<String>> filterColumns(SecurityContext securityContext, String catalogName, Map<SchemaTableName, Set<String>> tableColumns)
+    {
+        requireNonNull(securityContext, "securityContext is null");
+        requireNonNull(catalogName, "catalogName is null");
+        requireNonNull(tableColumns, "tableColumns is null");
+
+        Set<SchemaTableName> filteredTables = filterTables(securityContext, catalogName, tableColumns.keySet());
+        if (!filteredTables.equals(tableColumns.keySet())) {
```
[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/security/AccessControlManager.java:647]: i would expect equals to invoke contains only if sets are equal, and there is no second call then for given element

**Pull request merged**